File: shuar-scala/src/main/scala/shuar/Shuar.scala
package shuar

import shuar.Shuar.close

import scala.util.Try

trait Source[I]:
  def open(): Iterator[I]
end Source

trait Mapper[I, O]:
  def map(item: I): O
end Mapper

trait Reducer[O, R]:
  def reduce(items: Iterator[O]): R
end Reducer

trait Shuar[I, O, R](source: Source[I], mapper: Mapper[I, O], reducer: Reducer[O, R]):
  def mapReduce(): R =
    try {
      reducer.reduce(source.open().map(mapper.map))
    } finally {
      close(source, mapper, reducer)
    }
end Shuar
object Shuar:
  private def close(objs: Any*): Unit =
    objs
      .filter(_.isInstanceOf[AutoCloseable])
      .map(_.asInstanceOf[AutoCloseable])
      .foreach(c => Try(c.close()))
end Shuar

File: shuar-java/src/main/java/shuar/Shuar.java
package shuar;

import shuar.util.Utils;

import java.util.stream.Stream;

public class Shuar<I, O, R> {

    public interface Source<I> {
        Stream<I> open();
    }

    public interface Mapper<I, O> {
        O map(I item);
    }

    public interface Reducer<O, R> {
        R reduce(Stream<O> items);
    }

    private final Source<I> source;
    private final Mapper<I, O> mapper;
    private final Reducer<O, R> reducer;

    public Shuar(Source<I> source, Mapper<I, O> mapper, Reducer<O, R> reducer) {
        this.source = source;
        this.mapper = mapper;
        this.reducer = reducer;
    }

    public R mapReduce() {
        try {
            final var incomingItems = source.open();
            final var outgoingItems = incomingItems.map(mapper::map);
            return reducer.reduce(outgoingItems);
        } finally {
            close();
        }
    }

    private void close() {
        Stream.of(source, mapper, reducer)
                .filter(component -> component instanceof AutoCloseable)
                .map(component -> (AutoCloseable) component)
                .forEach(component -> Utils.runUnchecked(component::close));
    }
}
